### DISCUSSION TITLE
opposite of bootstrap?

### DISCUSSION THREAD

**Discussion by airtonix** [fluxcd/flux2#3119] [Q&A] [answered]
looks like we have to cobble together a whole bunch of logic to do the opposite of bootstrap? 

I'm trying to test a cold start, and whenever I discover some error in my manifests I need to do a complete reset so I'm having to basically : 

```shell
workstation>$ ssh themaster
themaster>$ sudo ./k3s-uninstall.sh
^d
workstation>$ k3sup ...
workstation>$ k3sup install --host ....
workstation>$ flux install ...
```

Surely there is a better way than having to completely destroy the k8s instance? 

I looked at `flux uninstall`, but it leaves all the pods etc still running.

**Comment by kingdonb** [comment, marked as answer]
That is correct about the behavior of `flux uninstall` – it is designed to remove Flux from control without disturbing any deployed resources, it would be a disaster if someone tried Flux as an experiment, then typed `flux uninstall` after trying it out, deciding it wasn't for them, and didn't realize it was going to wipe out all their deployed resources – so Flux won't do any of that.

Flux will only delete workloads if you enable `spec.prune` or Garbage Collection as it called in the docs. But when you use `flux uninstall`, the Flux controllers are stopped first, the workloads are suspended with no Flux controllers running, and then the CRDs are forcibly cleaned up, which leaves all the workloads intact. So if you want Flux to clean up all the workloads, you can delete the Flux Kustomizations after ensuring that `spec.prune` is `true`. Wait for those deletes to finish Terminating, and finally run `flux uninstall`. That should remove any trace that Flux was ever there, assuming you set up namespaces and you didn't do anything else outside of Flux, then the namespaces are also pruned and anything in them, with any cluster-wide resources.

Where you can still get in trouble is if you use other operators that also use finalizers, or any other operator that creates resources which are not marked with an ownerRef for cascasing deletion. I can't think of any specific example where this could happen, but in theory and in practice there are probably some. Maybe now you've deleted the operator of some other resource that used a finalizer, and now there is no controller to clean it up! You will have the same issue we described before with the finalizers.

Flux doesn't respect `dependsOn` order in reverse for deletion events (not yet, anyway):

* https://github.com/fluxcd/kustomize-controller/issues/301

But if you are going for realism in your "cold start" disaster recovery test, since this issue of leftover resources can impact from many ways (a job that creates namespaces, or something else in your stack with an imperative workflow rather than declarative), the best way is indeed probably to destroy the k8s instance and start a completely new one over again.
